src/dataworker/Dataworker.ts

```diff
@@ -42,68 +37,72 @@ export class Dataworker {
+            const depositUnfilledAmount = fill.amount.sub(fill.totalFilledAmount);
+            const depositKey = `${originChainId}+${fill.depositId}`;
+            assign(
+              unfilledDepositsForOriginChain,
+              [depositKey],
+              [
+                {
+                  unfilledAmount: depositUnfilledAmount,
+                  deposit: matchedDeposit,
+                  // A first partial fill for a deposit is characterized by one whose total filled amount post-fill
+                  // is equal to the amount sent in the fill, and where the fill amount is greater than zero.
+                  hasFirstPartialFill: fill.fillAmount.eq(fill.totalFilledAmount) && fill.fillAmount.gt(toBN(0)),
+                },
+              ]
+            );
+          } else {
+            this.logger.debug({
+              at: "Dataworker",
+              message: `Could not find deposit for fill on origin client`,
+              fill,
+            });
           }
-
-          return false; // No deposit matched, this fill is invalid.
-        });
-
-        this.logger.debug({
-          at: "Dataworker",
-          message: `Found ${validFillsOnDestinationChain.length} fills on destination ${destinationChainId} matching origin ${originChainId}`,
-          originChainId,
-          destinationChainId,
         });
-        validFillsOnDestinationChain.forEach((fill) =>
-          assign(fillsToRefund, [fill.repaymentChainId, fill.relayer], [fill])
-        );
       }
     }
 
+    // For each deposit with a matched fill, figure out the unfilled amount that we need to slow relay. We will filter
+    // out any deposits that are fully filled, or any deposits that were already slow relayed in a previous epoch.
+    const unfilledDeposits = Object.keys(unfilledDepositsForOriginChain)
+      .map((_originChainIdPlusDepositId: string): UnfilledDeposit => {
+        const _unfilledDeposits = unfilledDepositsForOriginChain[_originChainIdPlusDepositId];
+
+        // Remove deposits with no matched fills.
+        if (_unfilledDeposits.length === 0) return { unfilledAmount: toBN(0), deposit: undefined };
+        // Remove deposits where there isn't a fill with fillAmount == totalFilledAmount && fillAmount > 0. This ensures
+        // that we'll only be slow relaying deposits where the first fill
+        // (i.e. the one with fillAmount == totalFilledAmount) is in this epoch.
+        if (
+          !_unfilledDeposits.some((_unfilledDeposit: UnfilledDeposit) => _unfilledDeposit.hasFirstPartialFill === true)
+        )
+          return { unfilledAmount: toBN(0), deposit: undefined };
+        // Take the smallest unfilled amount since each fill can only decrease the unfilled amount.
+        _unfilledDeposits.sort((unfilledDepositA, unfilledDepositB) =>
+          unfilledDepositA.unfilledAmount.gt(unfilledDepositB.unfilledAmount) ? 1 : -1
```

Comment:
nit: in general, prefer sorting functions that don't depend on the ordering of the inputs. `cmp` returns 0 if they are equal: https://www.npmjs.com/package/bn.js. This can be possible I think if there are 0-sized fills.

```suggestion
          unfilledDepositA.unfilledAmount.cmp(unfilledDepositB.unfilledAmount)
```

Reply:
So it looks like ethers' `BigNumber` library doesn't expose `cmp`, would it be equivalent to just add the additional return value of 0 if they are equal?

Reply:
Yep!